app/models/user.rb

```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  before_save{self.email = email.downcase}
```

Comment:
hàm này dùng để làm gì? 
a dùng `downcase!` thì khác gì vs `downcase`

Reply:
before_save{self.email = email.downcase} : Trước khi lưu email của user nhập vào thì chuyển các kí tự trong chuỗi này thành chữ thường.
downcase : không thay đổi giá trị của chuổi nhập vào
downcase! : trả về giá trị nil nếu như xâu kí tự không có gì thay đổi

Reply:
Thế giá trị của chuỗi ban đầu ntn nếu dùng 2 hàm này em

Reply:
Giá trị ban đầu có chứa các kí tự viết hoa thì dowcase sẽ đưa về kí tự thường hết downcase! củng vậy.
Còn nếu giá trị ban đầu toàn kí tự viết thường thì downcase trả về y như giá trị đầu, còn dowcase! sẽ nil.

Reply:
Tìm hiểu lại chỗ này giúp anh nhé
Anh đang hỏi chuỗi gốc của nó không phải nói giá trị trả về

Reply:
Giá trị ban đầu có chứa các kí tự hoa thì dùng cả 2 cách đều đc.
Còn chứa toàn kí tự viết thường thì nên dùng downcase.
Trường hợp chưa biết giá trị người dùng nhập vào ntn thì nên dùng downcase.
Mong anh bổ sung thêm cho em :))

Reply:
email.downcase và email.downcase!
1. Giá trị trả về khác nhau như thế nào?
2.  Chuỗi gốc email ở đây khác nhau như thế nào?

Reply:
email.dowcase chỉ trả về giá trị nhưng ko làm thay đổi chuổi gốc
còn email.dowcase! vẫn trả về giá trị và làm thay đổi chuổi gốc.
ví dụ . a = "Abc" 
- a.dowcase = "abc", nhưng biến a = "Abc", không thay đổi a
- a.dowcase! = "abc", a ="abc", giá trị a đã bị thay đổi.